### PR TITLE
Spike-hancement: subcommands

### PR DESCRIPTION
**Pull request by bradsbrown**
Playing around with an idea @rbcasperson raised, to support a `jiratool` master command with subcommands under it. This PR maintains the "old" specific commands as well.

If we like this, I'll take any feedback you have and whip it into shape, or we can take it as-is plus docs updates...or if the feedback is this isn't useful, I can drop it.

### CODE REVIEW

File: jiratools/make_and_link.py
Comment (brolewis) [inline jiratools/make_and_link.py]: I know @sdutton-personal has found weird edge cases where Jira usually would return one data type but under certain conditions would return another. Is this a place where we should "guard" with jgt_common.list_from?

File: jiratools/helpers.py
Comment (brolewis) [inline jiratools/helpers.py]: This name just doesn't quite jive with me. Is it ha helper for finding a Jira or does it help find a JiraHelper-like thing.  What if maybe the name was changed to something like `get_issue_by_id`?

File: jiratools/helpers.py
Comment (brolewis) [inline jiratools/helpers.py]: Curious why this function requires `client` whereas the two below don't?

File: jiratools/example_config.py
Comment (brolewis) [inline jiratools/example_config.py]: Since `args.install` is the only item used form `args`, I'd be inclined to just pass that into `cli_example_config` instead of passing around the full `args` object.

File: jiratools/comment.py
Comment (brolewis) [inline jiratools/comment.py:14]: Is this debug code or needed to ensure a new line is present for the `print` functions below? If the later, can we add a comment so I don't see this in 3 months and ask this same dumb question again?

File: jiratools/helpers.py
Comment (brolewis) [reply jiratools/helpers.py]: I do see where it prevents creating a second `client` instance in `cli_make_linked`, but I'd be inclined to move it to an optional parameter.

File: .travis.yml
Comment (brolewis) [inline .travis.yml]: I'm not a huge fan of the ` | true` and would probably rather see the scripts themselves handle a case that would warrant this code.

File: jiratools/utils.py
Comment (brolewis) [inline jiratools/utils.py]: Is there a reason to set a `status` here? It doesn't look like that's being set anywhere else in the code.

File: jiratools/utils.py
Comment (brolewis) [inline jiratools/utils.py:22]: I don't _think_ we need a space here? But it might be a class-level thing

File: jiratools/example_config.py
Comment (brolewis) [reply jiratools/example_config.py]: Whatever the group decides, this would need to be consistent across all functions that accept `args`.

**Review by brolewis — COMMENTED** [review]
I don't think anything is a deal breaker and would be fine accepting this code as-is.

File: jiratools/utils.py
Comment (bradsbrown) [reply jiratools/utils.py:22]: It is a class-level requirement of our linting standards:
`D204 | 1 blank line required after class docstring`

File: jiratools/example_config.py
Comment (bradsbrown) [reply jiratools/example_config.py]: Since these are passed to cli-specific helper functions throughout, I decided to pass the full `args` object to each to allow us simpler future flexibility if we find a reason to tweak our args, but I'm open to further "votes" to refactor, if that's what we want.

File: jiratools/make_and_link.py
Comment (bradsbrown) [reply jiratools/make_and_link.py]: In this case, I don't *think* it should be, but if @sdutton-personal has input to affect my assumptions, I'm all ears.

**Review by brolewis — APPROVED** [review]
Talked with @bradsbrown and am OK with this PR as-is. The `| true` can be a later improvement and the `list_from` can be done if we ever hit a problem with it not being present.

File: jiratools/make_and_link.py
Comment (dgou) [reply jiratools/make_and_link.py]: Or we fix it as it breaks. Given the lack of reply by @sdutton-personal (who is on vacation and hopefully won't be disrupted by this mention), we can go forward as is.

File: .travis.yml
Comment (dgou) [reply .travis.yml]: Looking at this only in the tiny lens of the git diff, it's not immediately obvious why this is needed _here_.
Having now read through all the changes, I really don't like this... esp when it seems as if there would be a nice way to express the intention more clearly (see comments that follow)

File: bin/ensure_config.sh
Comment (dgou) [inline bin/ensure_config.sh]: So referencing the comment above, my guess is that this command could fail because a config is already installed.
In that case, I agree with @brolewis that this script should handle creating the correct return/exit status.
Maybe if `jira-example-config` had an `--install-if-missing` or some such, then this script wouldn't be needed and travis could call that directly? thinking out loud.

File: jiratools/__init__.py
Comment (dgou) [inline jiratools/__init__.py]: Not sure, but I think `Linke` shouldn't have the trailing `e`?

File: jiratools/__init__.py
Comment (dgou) [inline jiratools/__init__.py]: `Place` seems an odd verb here....??

File: jiratools/assignee.py
Comment (dgou) [inline jiratools/assignee.py]: Could be FF, but it'd be nice to include the user and JIRA id in the error message.
From a onesy-twosy on the command line, yeah, you'll "know" which user and JIRA, but if you are scripting this (even from the command line) and iterating over a bunch of JIRAs reassigning them (say because, cough, a coworker left), you won't know which one the failure refers to.

File: jiratools/comment.py
Comment (dgou) [inline jiratools/comment.py]: :+1: for showing the JIRA id here!

File: jiratools/example_config.py
Comment (dgou) [reply jiratools/example_config.py]: I"m on the edge of that coin, but I come down just slightly on the side of passing in the full `args`.

File: jiratools/example_config.py
Comment (dgou) [inline jiratools/example_config.py:24]: aha! here is where the travis travesity of working around the `--install` error happens.
Having thought about it for all of 10 minutes, I think having a default dry-run is the right thing, having a 'install or fail' is good for human use, _and_ having an 'install if needed' option is "just-right" for CICD situations. (In most cases the file will exist, but if the server is reset, a new server is started up, etc. it'll "just work").

File: pyproject.toml
Comment (dgou) [inline pyproject.toml:29]: Comment for future PRs. I'm not sure what happened here, but the new list of "existing" commands is not in the same order as the olde list which made it really annoying to make sure nothing got dropped. Had the new list been in alpha order, that've been OK too, but it seems to be randomized?

**Review by dgou — CHANGES_REQUESTED** [review]
Back to Brad for some discussion, potential tweaks, comments, feedback, clarifications, etc.

File: jiratools/__init__.py
Comment (ecedmondson) [reply jiratools/__init__.py]: Perform? Perhaps?

File: pyproject.toml
Comment (ecedmondson) [reply pyproject.toml:29]: I think the list is in the order that commands were added to the package

File: jiratools/example_config.py
Comment (ecedmondson) [reply jiratools/example_config.py]: I'm pro passing the full args fwiw

File: .travis.yml
Comment (ecedmondson) [reply .travis.yml]: I think that this was potentially resolved in the jgt slack development channel, but I think I'm going to hold off on providing the third approval until I double check.

**Review by ecedmondson — COMMENTED** [review]
Conditional approval until I check something. Ping me if I'm wrong.

File: jiratools/make_and_link.py
Comment (sdutton-personal) [reply jiratools/make_and_link.py]: Yeah, I agree, we can just fix it, if it becomes a problem.

File: jiratools/example_config.py
Comment (sdutton-personal) [reply jiratools/example_config.py]: I agree with Lewis on this FWIW.  I get that its easier in the future if the args is already there, but only passing what you need makes it more clear what is _currently_ being used IMO.

File: jiratools/example_config.py
Comment (dgou) [reply jiratools/example_config.py:24]: I think this is where some additional discussion may be appropriate?

**Review by brolewis — APPROVED** [review]

File: jiratools/__init__.py
Comment (rbcasperson) [inline jiratools/__init__.py:13]: Why all the `cli_` prefixes? Can't we assume everything is for the CLI since this is a CLI tool?

**Review by rbcasperson — COMMENTED** [review]
I'm slowing getting back some context here. I think the code changes look good to me, but can you first update the README and any other docs/tests so I can start there with what the changes are? I'm fine with keeping the legacy commands for now, but let's document the "correct" ones.

File: jiratools/__init__.py
Comment (ecedmondson) [reply jiratools/__init__.py:13]: I know of at least one repo that used some of these functions for non-clie tools...but it's an enterprise repo that might not be supported for much longer due to things outside of the maintainer's control.

File: jiratools/__init__.py
Comment (ecedmondson) [inline jiratools/__init__.py:48]: `Linke` reminds me of old english for some reason

File: bin/ensure_config.sh
Comment (ecedmondson) [inline bin/ensure_config.sh:3]: 👍

**Review by ecedmondson — APPROVED** [review]
I had to go back and look at how the subcommands were implemented to get context on this, and then further, get context on the individual args. Definitely think this is the better option for args-y stuff. Looks good.

**Review by sdutton-personal — APPROVED** [review]
I have no objections.  I have been trying to transition the repo that uses this off any ways due to it murdering the jira server.

**Comment by bradsbrown** [comment]
*FInally* got around to updating the README. I think that's all that was requested (no tests in this repo), right @rbcasperson?

**Comment by ecedmondson** [comment]
Haven't looked too deeply into this yet and won't have time until later today, but it looks like the PR check failed in Travis @rbcasperson and @bradsbrown

**Review by rbcasperson — APPROVED** [review]

**Pull request merged**
